### PR TITLE
incr.comp.: Avoid creating an edge to DepNode::Krate when generating debuginfo namespaces.

### LINKED ISSUE #38222: incr.comp.: Edge to Krate depnode when building regex in debug mode
Occurs with `regex 0.1.75 (ac9ced6)`. The Krate-edge leads to everything being re-compiled for any change. The edge *does not* occur when compiling with `--release`.

### PR DESCRIPTION
**Pull request by michaelwoerister**
r? @nikomatsakis 

Fixes #38222

### CODE REVIEW

File: src/librustc_trans/debuginfo/namespace.rs [hunk 1]
```diff
@@ -63,13 +63,21 @@ pub fn item_namespace(ccx: &CrateContext, def_id: DefId) -> DIScope {
...
-    let namespace_name = match def_key.disambiguated_data.data {
-        DefPathData::CrateRoot => ccx.tcx().crate_name(def_id.krate).as_str(),
-        data => data.as_interned_str()
+    let (namespace_name, span) = match def_key.disambiguated_data.data {
+        DefPathData::CrateRoot => {
+            // We don't provide the span for the crate root. Because of
+            // implementation reasons it is not possible to access this
+            // information without introducing a dep-tracking edge to
+            // DepNode::Krate, which essentially breaks incr. comp.
+            (ccx.tcx().crate_name(def_id.krate).as_str(), DUMMY_SP)
```
Comment (nikomatsakis) [inline src/librustc_trans/debuginfo/namespace.rs]: Hmm. I guess you decided that doing the refactoring to break out the `span` was too difficult? Seems like we could add a `ccx.tcx().map.crate_span()` helper that just ignores the read-edge easily enough as well, no?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_trans/debuginfo/namespace.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc_trans/debuginfo/namespace.rs]: Well I would point it at `DepNode::Hir(CRATE_DEF_ID)`, I think.

Note that DepNode::Krate and the Hir node for the crate def-id are distinct nodes.

**Comment by michaelwoerister** [comment]
Pushed a cleaner fix for this. Reading a the crate's span or attribute will now add an edge to `DepNode::Hir(CRATE_DEF_INDEX)` instead of to the all encompassing `DepNode::Krate`.

**Pull request merged**